### PR TITLE
Introduce VgicConfig structure for initializing Vgic

### PR DESCRIPTION
**Pull request by NunoDasNeves**
This series introduces a structure and a couple of helpers which are used to initialize the Vgic.
- Instead of computing GIC redist/msi regs everywhere, only compute it in devices::gic::create_vgic_config()
- For tests, use the config created by hypervisor::arch::aarch64::gic::create_test_vgic_config()
- Passing the config from a higher level (configure_system()) makes the code more flexible - GIC can be placed anywhere and the code only has to be modified in one place

### CODE REVIEW

File: hypervisor/src/vm.rs
Comment (sboeuf) [inline hypervisor/src/vm.rs:276]: Well I guess you don't need that given you've replaced all the parameters with `VgicConfig`.

**Review by sboeuf — APPROVED** [review]

**Comment by rbradford** [comment]
I think the first lines of the commit messages should be more like "devices: gic"

File: devices/src/gic.rs
Comment (rbradford) [inline devices/src/gic.rs:30]: Woulndn't `VgicConfig::new()` be more appropriate

File: hypervisor/src/arch/aarch64/gic.rs
Comment (rbradford) [inline hypervisor/src/arch/aarch64/gic.rs:31]: I don't think this is correct. You should put it inside a `mod tests` block

**Review by rbradford — CHANGES_REQUESTED** [review]
I think there needs to be some changes before this lands.

File: devices/src/gic.rs
Comment (NunoDasNeves) [reply devices/src/gic.rs:30]: I can't implement VgicConfig::new() here, since it's in a different crate from where VgicConfig is defined

File: hypervisor/src/vm.rs
Comment (NunoDasNeves) [reply hypervisor/src/vm.rs:276]: Oops! Left over from a previous iteration of this patch...

File: devices/src/gic.rs
Comment (rbradford) [reply devices/src/gic.rs:30]: It would make make most sense for them to be together, no?

Comment (NunoDasNeves) [reply devices/src/gic.rs:30]: VgicConfig has to be defined in the hypervisor crate (which makes sense - along with the Vgic trait).
But the hypervisor crate can't access arch::layout or devices::gic so I put create_vgic_config here...

I tried making the hypervisor crate depend on the arch crate but it creates a cyclic dependency.
Making devices crate depend on hypervisor crate also creates a cycle...

I'm not sure how to put VgicConfig and its initialization together and have them accessible from inside and outside the hypervisor crate!

Comment (rbradford) [inline devices/src/gic.rs:30]: Okay. Move this to `Gic::create_default_config()`

Comment (NunoDasNeves) [reply devices/src/gic.rs:30]: ok!

**Review by rbradford — APPROVED** [review]
Fine with me but @michael2012z  or @MrXinWang should review as aarch64 experts.

**Review by russell-islam — APPROVED** [review]

**Review by michael2012z — APPROVED** [review]
Good improvement!

The settings of vgic attributes were here and there. Now it's much easier to manage these configurations.

Thanks.

**Review by MrXinWang — APPROVED** [review]

**Pull request merged**
